### PR TITLE
initial backend music controller work and user preferences updates

### PR DESCRIPTION
**Pull request by hannahbergam**
This is step one of three PRs to implement the new Mute Music functionality for craft and star wars levels. Currently, the background music plays automatically (and somewhat loudly) when the levels are opened, and there is no way to turn off the music aside from the user muting the tab or their computer, thus losing all additional sounds.

The overall goal: add a mute music button to the instructions panel (design pending user group testing) on these levels, which then persists across levels for both signed in and signed out users.

This PR deals specifically with adding a serialized attribute to user.rb, and a userPreference for muteMusic, so that once a user decides to mute the background music, it is saved on their account and applied thereafter until they decide to 'unmute'. We will use a cookie to implement this functionality for signed-out users. Currently the logic operates as follows: if there is a muteMusic cookie or a user preference set, the background music will be muted. Otherwise: play all sounds (current implementation). 

## Links

- spec: []()
- jira ticket: [] https://codedotorg.atlassian.net/browse/LP-2230

## Testing story

This PR adds tests for get and post requests in the user controller test file, following the pattern created by 'using_text_mode'.

That said, this is only the back end set up for the work, and following PRs will include UI and integration testing.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -58,6 +58,11 @@ def get_display_theme
     render json: {display_theme: @user&.display_theme}
   end
 
+  # GET /api/v1/users/<user_id>/mute_music
+  def get_mute_music
+    render json: {mute_music: !!@user.mute_music}
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/users_controller.rb]: nit: use `@user&`

Comment (bethanyaconnor) [inline dashboard/app/controllers/api/v1/users_controller.rb]: nit: use `@user&`

File: dashboard/config/routes.rb [hunk 2]
```diff
@@ -768,8 +768,10 @@
       concerns :section_api_routes
       post 'users/:user_id/using_text_mode', to: 'users#post_using_text_mode'
       post 'users/:user_id/display_theme', to: 'users#update_display_theme'
+      post 'users/:user_id/mute_music', to: 'users#post_mute_music'
```
Comment (bethanyaconnor) [inline dashboard/config/routes.rb:780]: For my knowledge: does this just work by some Rails magic? I don't see a controller action defined?

Comment (bethanyaconnor) [inline dashboard/config/routes.rb:780]: For my knowledge: does this just work by some Rails magic? I don't see a controller action defined?

File: apps/src/MusicController.js [hunk 3]
```diff
@@ -171,7 +175,7 @@ MusicController.prototype.preload = function() {
  */
 MusicController.prototype.play = function(trackName) {
   debug('play ' + trackName);
-  if (!this.audioPlayer_) {
+  if (!this.audioPlayer_ || this.muteMusic_) {
```
Comment (bethanyaconnor) [inline apps/src/MusicController.js:178]: Does this mean that this setting will mute the sound now? (I understand we don't yet have a way for user to actually set this setting, but if we set it manually in the DB or something.)

I think you answered this question at checkin but this was the comment I'd half typed out 😅

Comment (bethanyaconnor) [inline apps/src/MusicController.js:178]: Does this mean that this setting will mute the sound now? (I understand we don't yet have a way for user to actually set this setting, but if we set it manually in the DB or something.)

I think you answered this question at checkin but this was the comment I'd half typed out 😅

**Review by bethanyaconnor — APPROVED** [review]
Looks good!

File: dashboard/config/routes.rb (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/config/routes.rb:780]: It looks like `using_text_mode` does have a controller action: https://github.com/code-dot-org/code-dot-org/blob/d246587fb1c6a0422fdc9dd1bdf322af18160388/dashboard/app/controllers/api/v1/users_controller.rb#L102. So I'm still very curious :)

File: dashboard/app/controllers/api/v1/users_controller.rb (hunk 1 above)
Comment (hannahbergam) [reply dashboard/app/controllers/api/v1/users_controller.rb]: 🙌🏼

Comment (hannahbergam) [reply dashboard/app/controllers/api/v1/users_controller.rb]: Thanks for this- I updated the using_text_mode instances too.

File: dashboard/config/routes.rb (hunk 2 above)
Comment (hannahbergam) [reply dashboard/config/routes.rb:780]: 🤦🏻‍♀️🤦🏻‍♀️🤦🏻‍♀️ Fixed and updated!

Comment (hannahbergam) [reply dashboard/config/routes.rb:780]: (Still facepalming) - thank you for noticing this, it's added/fixed now!

File: apps/src/MusicController.js (hunk 3 above)
Comment (hannahbergam) [reply apps/src/MusicController.js:178]: Haha yes! Well almost- this setting will mute the background music, but not the actual sounds the little sprites make when they move around. There will be a way for the user to set it soon:)

**Review by bethanyaconnor — APPROVED** [review]
🎉 🔇

**Pull request merged**